lib/staccato/v4/event.rb

```diff
@@ -1,76 +1,145 @@
+          extend Forwardable
+
+          def_delegators :@options, :[], :[]=
+
+          @recommended_params = []
+          fields = defined?(klass::FIELDS) ? klass::FIELDS : []
+          fields.each do |name|
+            param name
           end
-        end
 
-        def name
-          self.class.event_name
+          # In order for user activity to display in standard reports like
+          # Realtime, engagement_time_msec and session_id must be supplied as
+          # part of the params for an event.
+          param :engagement_time_msec
+          param :session_id
+
+          # The time to associate with the event, as a Unix timestamp in
+          # microseconds.  This should only be set to record events that
+          # happened in the past.
+          #
+          # Overrides the tracker timestamp_micros.
+          param :timestamp_micros
         end
+
+        events[klass.event_name] = klass
       end
 
-      events[event.event_name] = event
     end
 
-    def initialize(tracker, options = {})
-      self.tracker = tracker
-      self.options = Staccato::OptionSet.new(options)
+    module ClassMethods
+
+      attr_reader :recommended_params
+
+      def event_name
+        @event_name ||= name
+          .dup
+          .split('::')
+          .last
+          .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
+          .gsub(/([a-z\d])([A-Z])/, '\1_\2')
+          .tr('-', '_')
+          .downcase
+          .to_sym
+      end
+
+      def param(name)
+        # TODO: prevent assignment of restricted params
+        name = name.to_sym
+        recommended_params << name
+        def_delegator :@options, name
+      end
+
     end
 
-    # return the fields for this hit type
-    # @return [Array] the fields
-    def fields
-      self.class::FIELDS
+    def initialize(**options)
+      @options = Staccato::OptionSet.new(options)
+    end
+
+    def name
+      self.class.event_name
     end
 
     def as_json
-      {
-        name: name,
-        params: event_params
-      }
+      Staccato::V4.encode_params_json({name: name, params: params})
     end
 
-    private
+    def ==(other) other.class == self.class && other.as_json == as_json end
+    alias_method :eql?, :==
+
+    def hash; [self.class, as_json].hash end
+
+    # The event parameters to be sent, including custom parameters
+    # @return [Hash{Symbol=>Numeric,String,Array[Hash{Symbol=>Numeric,String}]}]
+    def params
+      Staccato::V4.encode_params_json options.to_h.compact
+    end
 
-    def event_params
-      fields.map { |field|
+    # The recommended parameter values for this event
+    # @return (see params)
+    def recommended_params
+      self.class.recommended_params.map {|field|
         [field, options[field]] unless options[field].nil?
       }.compact.to_h
     end
 
-    # include BooleanHelpers
-    # adds convert_booleans()
+    # The custom event parameters to be sent for this event
+    # @return (see params)
+    def custom_params
```

Comment:
This doesn't appear to be used yet?

Reply:
It's for the end-user to use as they see fit, and for documentation.  All event params (both "recommended" and "custom") are stored on `params` and encoded by `as_json`.

Reply:
I'd like to hold off on this for now until there's a use for it before we add it to the public API.